gap_darwin.go

```diff
@@ -127,10 +132,11 @@ func (a *Adapter) Connect(address Addresser, params ConnectionParams) (*Device,
 		d.delegate = &peripheralDelegate{d: d}
 		p.SetDelegate(d.delegate)
 
-		a.connectHandler(nil, true)
+		a.connectHandler(adr, true)
 
 		return d, nil
-	case <-time.NewTimer(10 * time.Second).C:
+	case <-time.NewTimer(timeout).C:
+		a.cm.CancelConnect(prphs[0]) // we need to cancel the connection if we have timed out ourselves
```

Comment:
Connections can take longer than the default 10 seconds to establish in some cases. This makes sure that when the user is indicated a connection timeout, the connection doesn't actually continue to be established in the background.

Reply:
Good idea.
It _might_ be a good idea to wait until the device is disconnected (`didDisconnectPeripheral`):
https://developer.apple.com/documentation/corebluetooth/cbcentralmanager/1518952-cancelperipheralconnection

Reply:
Okay, so I have modified the `DidDisconnectPeripheral` handler to also interact with the `adapter.connectMap` in the same way as the `DidConnectPeripheral` handler. This way, the `prphCh` in the `adapter.Connect` function will get a response on connect and disconnect, and we can check the peripheral state when it arrives. 

This does mean that the switch case has to be wrapped so that when the timer expires, we continue waiting for this disconnection event to trigger. This poses a concern if nothing ever shows up, and I wonder if we should start some form of bail-out timer just in case.